keps/sig-scheduling/20200612-simplified-topology-manager.md

```diff
@@ -0,0 +1,272 @@
+- [Use cases](#use-cases)
+- [Known limitations](#known-limitations)
+- [Test plans](#test-plans)
+- [Graduation criteria](#graduation-criteria)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature enablement and rollback](#feature-enablement-and-rollback)
+- [Implementation history](#implementation-history)
+<!-- /toc -->
+
+# Summary
+
+This document describes behaviour of the Kubernetes Scheduler which takes worker node topology into account .
+
+# Motivation
+
+After Topology Manager was introduced the problem of launching pod in the
+cluster where worker nodes have different NUMA topology and different amount
+of resources in that topology became actual. Pod could be scheduled on the node
+where total amount of resources are enough, but resource distribution could not
+satisfy the appropriate Topology policy. In this case the pod failed to start. Much
+better behaviour for scheduler would be to select appropriate node where kubelet admit
+handlers may pass.
+
+
+## Goals
+
+- Make scheduling process more precise when we have NUMA topology on the
+worker node.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a specific node topology manager policy
+- This Proposal requires exposing NUMA topology information. This KEP doesn't
+describe how to expose all necessary information it just declare what kind of
+information is necessary.
+
+# Proposal
+
+Kube-scheduler built-in plugin will be added to the main tree. This plugin
+implements a simplified version of Topology Manager and hence is different from original topology manager algorithm.
+Plugin would be disabled by default and when enabled would check for the ability to run pod only in case of single-numa-node policy on the
+node, since it is the most strict policy, it implies that the launch on the node with
+other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
+Proposed plugin will use [CRD][1] to identify which topology policy is enabled on the node.
+To work, this plugin requires topology information of the available resource on the worker nodes.
+
+## Topology format
+
+Available resources with topology of the node should be stored in CRD. Format of the topology described
+[in this document][1].
+
+The daemon which runs outside of the kubelet will collect all necessary information on running pods, based on allocatable resources of the node and consumed resources by pods it will provide available resources in CRD, where one CRD instance represents one worker node. The name of the CRD instance is the name of the worker node.
+
+## CRD API
+
+Code responsible for CRD API will be placed in [node feature discovery repository][2]. To use it in the kube-scheduler kubernetes should keep NFD in its staging directory. To reduce the size of dependency just [code for working with CRD][3] could be placed in staging directory.
```

Comment:
this is awkward, and i am aware of no similar precedent.  this means the API is versioned with k8s and not with nfd.

Reply:
we would want nfd to clarify when it cuts releases to make this really work.

Reply:
@marquiz @zvonkok Can you please provide us information about the release cadence of NFD? Are releases cut on a regular cadence or ad hoc?

Reply:
So far we've released on an ad hoc / as-needed basis

Reply:
I wouldn't be too excited about a cyclic dep between nfd and k8s 🤔 
'

Reply:
I asked this is b/c: it seems you're targeting to have the CRD placed in k/k staging (in the "Graduation criteria" section). In that case, it's a built-in API, not CRD. (technically CRD and in-tree API are not treated as the same thing). So I want to ensure everyone is on the same page.

BTW: I recalled that when RuntimeClass was alpha (in 1.13?), it was introduced as a CRD, and maintained in an external repo. And later after (1.14?), it was moved to k/k when it's beta. Probably we should refer to that, and make make the roadmap of the API clear to everyone.

For an alpha feature (disabled by default), I think it's sort of ok to vendor an external repo from k/k (scheduler and kubelet), as long as there is no cyclic dependency. While for a beta feature, vendoring an external CRD doesn't seem to have a precedent (let me know if there is).

Reply:
@derekwaynecarr, what if we will place CRD API code in kubernetes/kubernetes repository, somewhere in staging/src/k8s.io/api/. Since NFD already depends on k8s, that CRD API code could be used both in kube-scheduler plugin and in NFD.

Reply:
> @swatisehgal is the CRD API available in NFD repo already? and is there any consensus on how to evolve the API? e.g., v1alphaX to be maintained in NFS repo, and when it's graduated to v1betaY, merge it back to k/k?

@Huang-Wei Work to enable NFD to gather topology based resource information is in progress, we are using an external repo for API definition for prototyping so it's not in NFD currently. Also, thanks for sharing  RuntimeClass as an example. Like you said we need a a proper roadmap for the API, we would like to get some input from @derekwaynecarr  on this to conform to the best know practices for a feature like this. From the comments so far given that vendoring in NFD is not appropriate due to cyclic dependency alternative options are:

1. having it in k/k so both NFD and scheduler can refer to it

Reply:
Either 1 or 2 would work. For 1, we just need to ensure scheduler to only depend on the API (and codegen) pkgs in NFD, which shouldn't too hard to achieve.